File: agent/scc_agent/creds.py
"""Credential storage. On Windows uses DPAPI (LocalMachine scope); falls back to obfuscated
local file on non-Windows for development."""
from __future__ import annotations

import base64
import json
import os
import pathlib
import sys
from typing import Optional

from scc_agent.config import CREDS_PATH, ensure_dirs

_DEV_KEY = b"dev-only-not-secure-key-32bytes!"


def _is_windows() -> bool:
    return sys.platform == "win32"


def _xor(data: bytes, key: bytes) -> bytes:
    return bytes(b ^ key[i % len(key)] for i, b in enumerate(data))


def save(key_id: str, secret: str, path: pathlib.Path | None = None) -> None:
    ensure_dirs()
    target = path or CREDS_PATH
    blob = json.dumps({"key_id": key_id, "secret": secret}).encode()
    if _is_windows():
        import win32crypt  # type: ignore

        enc = win32crypt.CryptProtectData(blob, None, None, None, None, 0x4)  # LOCAL_MACHINE
    else:
        enc = b"DEV1" + base64.b64encode(_xor(blob, _DEV_KEY))
    target.write_bytes(enc)
    try:
        os.chmod(target, 0o600)
    except Exception:
        pass


def load(path: pathlib.Path | None = None) -> tuple[str, str]:
    target = path or CREDS_PATH
    enc = target.read_bytes()
    if _is_windows() and not enc.startswith(b"DEV1"):
        import win32crypt  # type: ignore

        _, dec = win32crypt.CryptUnprotectData(enc, None, None, None, 0)
    else:
        dec = _xor(base64.b64decode(enc[4:]), _DEV_KEY)
    obj = json.loads(dec)
    return obj["key_id"], obj["secret"]


def try_load(path: pathlib.Path | None = None) -> Optional[tuple[str, str]]:
    try:
        return load(path)
    except FileNotFoundError:
        return None

File: go_agent/internal/creds/creds.go
// Package creds stores the agent's API key. On Windows it uses DPAPI
// (LocalMachine scope); on other platforms it falls back to an obfuscated
// dev file. Mirrors scc_agent/creds.py exactly, including the "DEV1" prefix
// and XOR key, so dev cred files remain interchangeable.
package creds

import (
	"encoding/base64"
	"encoding/json"
	"os"

	"github.com/complete3tech/scc-agent/internal/config"
)

const (
	devKey    = "dev-only-not-secure-key-32bytes!"
	devPrefix = "DEV1"
)

type blob struct {
	KeyID  string `json:"key_id"`
	Secret string `json:"secret"`
}

// Save persists the key/secret pair to CredsPath (0600).
func Save(keyID, secret string) error {
	if err := config.EnsureDirs(); err != nil {
		return err
	}
	raw, err := json.Marshal(blob{KeyID: keyID, Secret: secret})
	if err != nil {
		return err
	}
	enc, err := protect(raw)
	if err != nil {
		return err
	}
	return os.WriteFile(config.CredsPath, enc, 0o600)
}

// Load returns (key_id, secret) from CredsPath.
func Load() (string, string, error) {
	enc, err := os.ReadFile(config.CredsPath)
	if err != nil {
		return "", "", err
	}
	dec, err := unprotect(enc)
	if err != nil {
		return "", "", err
	}
	var b blob
	if err := json.Unmarshal(dec, &b); err != nil {
		return "", "", err
	}
	return b.KeyID, b.Secret, nil
}

// TryLoad is Load but reports presence as a bool instead of an error.
func TryLoad() (string, string, bool) {
	k, s, err := Load()
	if err != nil {
		return "", "", false
	}
	return k, s, true
}

func xorBytes(data []byte) []byte {
	key := []byte(devKey)
	out := make([]byte, len(data))
	for i := range data {
		out[i] = data[i] ^ key[i%len(key)]
	}
	return out
}

func devEncode(b []byte) []byte {
	return append([]byte(devPrefix), []byte(base64.StdEncoding.EncodeToString(xorBytes(b)))...)
}

func devDecode(enc []byte) ([]byte, error) {
	raw, err := base64.StdEncoding.DecodeString(string(enc[len(devPrefix):]))
	if err != nil {
		return nil, err
	}
	return xorBytes(raw), nil
}
